### ISSUE TITLE
Question: what's the meaning of num_cpus? 

### ISSUE THREAD

**Issue by vycezhong** [ray-project/ray#15613]
Hi! I am reading docs and get confused with this part: 

> Tasks: When Ray starts on a machine, a number of Ray workers will be started automatically (1 per CPU by default). They will be used to execute tasks (like a process pool). If you execute 8 tasks with num_cpus=2, and total number of CPUs is 16 (ray.cluster_resources()[“CPU”] == 16), you will end up with 8 of your 16 workers idling.

It seems that specifying _num_cpus=2_ makes no difference here. Only half of  CPUs are working and each task only utilizes one CPU. Then what's the meaning of _num_cpus_?

**Comment by krfricke** [comment]
Hi @vycezhong, maybe the example is a bit hard to parse. Let me try to lay it our a little bit more clear.

The example assumes you have a cluster (or a single machine) with 16 CPUs.

You start Ray with no specific options. Ray thus starts a python worker process for each CPU, so 16 CPUs.

You can see these worker processes e.g. with `ps aux`. They are currently in idle state.

Then you start 8 tasks with `num_cpus=2`. This refers to the ray remote decorator, i.e. this: @ray.remote(num_cpus=2)`. This means each worker should use 2 CPUs.

Ray knows that the cluster has 16 CPUs available. If each task occupies 2 CPUs, it thus concludes it can only start up to 8 tasks concurrently. This means that the other 8 ray worker processes remain idle.

I hope this makes it clear. The important concept here is that Ray's `num_cpus` concept translates to processes running on the host OS. This is also the reason why you are able to start ray with more (virtual) CPUs than the host system has available: You can easily call `ray.init(num_cpus=100)` on a system with say 16 cores. This will start 100 processes on the machine. However, if you then proceed to start say 100 remote tasks, the OS takes care of allocating CPU resources to the tasks. If you use as many (or fewer) worker processes as the machine actually has, the OS doesn't have to switch between processes and can fully utilize all computation power in parallel.

Does this make sense?

**Comment by vycezhong** [comment]
@krfricke  Thank you so much for your detailed explanation. BTW, if `num_cpus=2` are given, I wonder whether the task can actually occupy more than 2 CPUs?

**Comment by krfricke** [comment]
If the actor starts subprocesses or threads this could happen. It's up to the user to prevent that from happening.

Per default, Ray sets `OMP_NUM_THREADS=1` to avoid this for numeric libraries such as numpy or pandas.

**Comment by vycezhong** [comment]
Got it. Thank you!
